ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/gcs.rs

```diff
@@ -0,0 +1,305 @@
+                    bucket: self.bucket_name.clone(),
+                    ..Default::default()
+                },
+                file_stream,
+                &UploadType::Simple(Media {
+                    name: filename.clone().into(),
+                    content_type: Borrowed(TAR_FILE_TYPE),
+                    content_length: None,
+                }),
+            )
+            .await
+        {
+            Ok(_) => {
+                self.update_metadata(chain_id, epoch).await?;
+                let snapshot_path_clone = snapshot_path.clone();
+                fs::remove_file(&tar_file)
+                    .and_then(|_| fs::remove_dir_all(snapshot_path_clone))
+                    .await
+                    .expect("Failed to clean up after db snapshot upload");
+            },
+            Err(err) => {
+                error!("Failed to upload snapshot: {}", err);
+            },
+        };
+
+        Ok(())
+    }
+
+    /// When fullnode is getting started, it will first restore its table info db by restoring most recent snapshot from gcs buckets.
+    /// Download the right snapshot based on epoch to a local file and then unzip it and write to the indexer async v2 db.
+    pub async fn restore_db_snapshot(
+        &self,
+        chain_id: u64,
+        metadata: BackupRestoreMetadata,
+        db_path: PathBuf,
+        base_path: PathBuf,
+    ) -> anyhow::Result<()> {
+        assert!(metadata.chain_id == chain_id, "Chain ID mismatch.");
+
+        let epoch = metadata.epoch;
+        let epoch_based_filename = generate_blob_name(epoch);
+
+        match self
+            .gcs_client
+            .download_streamed_object(
+                &GetObjectRequest {
+                    bucket: self.bucket_name.clone(),
+                    object: epoch_based_filename.clone(),
+                    ..Default::default()
+                },
+                &Range::default(),
+            )
+            .await
+        {
+            Ok(mut stream) => {
+                // Create a temporary file and write the stream to it directly
+                let temp_file_name = "snapshot.tar.gz";
+                let temp_file_path = base_path.join(temp_file_name);
+                let temp_file_path_clone = temp_file_path.clone();
+                let mut temp_file = File::create(&temp_file_path_clone).await?;
+                while let Some(chunk) = stream.next().await {
+                    match chunk {
+                        Ok(data) => temp_file.write_all(&data).await?,
+                        Err(e) => return Err(anyhow::Error::new(e)),
+                    }
+                }
+                temp_file.sync_all().await?;
+
+                // Spawn blocking a thread to synchronously unpack gzipped tar file without blocking the async thread
+                task::spawn_blocking(move || {
+                    unpack_tar_gz(&temp_file_path_clone, &db_path)
+                        .expect("Failed to unpack gzipped tar file");
```

Comment:
nit: I think it's better to panic the current task than in the blocking task runner thread but I don't know why (maybe you will see `restore_db_snapshot` in the stack trace instead of `closure`) :trollface: